device_specific/flash_the_ttgo_tdisplay

```diff
@@ -81,25 +93,183 @@ fi
+
+  sed -i.bak '/CONFIG_ESP32_DISABLE_BASIC_ROM_CONSOLE=/d' sdkconfig.defaults
+  echo "CONFIG_ESP32_DISABLE_BASIC_ROM_CONSOLE=y" >> sdkconfig.defaults
+
+  sed -i.bak '/CONFIG_SECURE_DISABLE_ROM_DL_MODE=/d' sdkconfig.defaults
+  echo "CONFIG_SECURE_DISABLE_ROM_DL_MODE=y" >> sdkconfig.defaults
+
+  sed -i.bak '/CONFIG_SECURE_BOOT_SIGNING_KEY=/d' sdkconfig.defaults
+  echo 'CONFIG_SECURE_BOOT_SIGNING_KEY="../../../jade-diy.pem"' >> sdkconfig.defaults
+
+  sed -i.bak '/CONFIG_SECURE_BOOT=/d' sdkconfig.defaults
+  echo "CONFIG_SECURE_BOOT=y" >> sdkconfig.defaults
+
+  sed -i.bak '/CONFIG_SECURE_FLASH_ENC_ENABLED=/d' sdkconfig.defaults
+  echo "CONFIG_SECURE_FLASH_ENC_ENABLED=y" >> sdkconfig.defaults
+
+  sed -i.bak '/CONFIG_SECURE_FLASH_ENCRYPTION_MODE_RELEASE=/d' sdkconfig.defaults
+  echo "CONFIG_SECURE_FLASH_ENCRYPTION_MODE_RELEASE=y" >> sdkconfig.defaults
+
+  sed -i.bak '/CONFIG_ESP32_REV_MIN_/d' sdkconfig.defaults # remove all REV_MIN entries
+  echo "CONFIG_ESP32_REV_MIN_3=y" >> sdkconfig.defaults
+
+  echo "Secure Boot configuration has been updated in sdkconfig.defaults... building bootloader"
+  idf.py bootloader
+fi
 rm sdkconfig.defaults.bak
 
+# workaround to prevent changed idf manifest hash to flag build as dirty see https://github.com/Blockstream/Jade/issues/98
+git restore dependencies.lock
+
 idf.py build
 
 [ "${CI:-false}" = true ] && echo "Exiting the script for CI runners." && exit 0
 
-while [ ! -c "${tty_device}" ]; do
-  read -srn1 -p "Connect your ${chosen_device} and PRESS ANY KEY to continue... " && echo
-done
-initial_tty_device_permissions="$(stat -c '%a' "${tty_device}")"
-if [ "${initial_tty_device_permissions:2}" -lt 6 ]; then
-  echo -e "\nElevating write permissions for ${chosen_device}"
-  sudo chmod o+rw "${tty_device}"
-  echo
+if [[ "$enable_secure_boot" == "u" ]]; then
+  echo "Upgrading secure boot enabled device... please confirm on screen"
+  while true; do
+    if [ -f ~/jade-diy.pem ] && [ -r ~/jade-diy.pem ]; then
+      echo "The signing key ~/jade-diy.pem exists and is readable."
+      break
+    else
+      clear
+      echo "Error: Signing key ~/jade-diy.pem missing. Please make sure that the used signing key ~/jade-diy.pem exists and is readable under the specified path."
+      echo "The script will continue once the key is available"
+      echo "(Ctrl + c to cancel)"
+      sleep 5
+    fi
+  done
+  ./jade_ota.py --noagent
+  exit 0
 fi
 
-idf.py flash
-
+if [[ "$enable_secure_boot" == "y" ]]; then
+  echo "Flashing bootloader..."
+  sleep 2
+  esptool.py --chip esp32 --before=default_reset --after=no_reset write_flash --flash_mode dio --flash_freq 40m --flash_size keep 0x1000 build/bootloader/bootloader.bin
+  echo "Flashing app..."
+  sleep 2
+  esptool.py --before default_reset --after no_reset --chip esp32 write_flash --flash_mode dio --flash_size keep --flash_freq 40m 0x9000 build/partition_table/partition-table.bin 0xe000 build/ota_data_initial.bin 0x10000 build/jade.bin
```

Comment:
Did you pull these from https://github.com/Blockstream/Jade/blob/1e5f9d163adfada1a83fb9b34e8435236d0e8a16/main/qemu/make-flash-img.sh#L13?

These memory addresses should be the same for all four devices, right?

Reply:
I took these from the output of the esptool.py after building the bootloader as suggested in the youtube tutorial here https://youtu.be/PeqP6oVnlIs?t=1231 (and removed port and baut rate as suggested).

Actually i am not sure if the addresses are the same for all devices, i would except that but i cant test it.